File: src/codes/no-undo-history/ts/rfc_command.ts
// コマンドインターフェース
interface Command {
  execute(): void;
  undo(): void;
}

// 受信者（対象クラス）
class TextEditor {
  private content = "";

  add(text: string) {
    this.content += text;
  }

  remove(length: number) {
    this.content = this.content.slice(0, -length);
  }

  getContent(): string {
    return this.content;
  }
}

// コマンド実装
class TypeCommand implements Command {
  constructor(private editor: TextEditor, private text: string) {}

  execute(): void {
    this.editor.add(this.text);
  }

  undo(): void {
    this.editor.remove(this.text.length);
  }
}

// 利用例
const editor = new TextEditor();
const history: Command[] = [];

const cmd1 = new TypeCommand(editor, "Hello ");
cmd1.execute();
history.push(cmd1);

const cmd2 = new TypeCommand(editor, "World");
cmd2.execute();
history.push(cmd2);

console.log(editor.getContent()); // Hello World

// Undo（巻き戻し）
const last = history.pop();
last?.undo();

console.log(editor.getContent()); // Hello

File: src/codes/no-undo-history/py/rfc_command.py
from abc import ABC, abstractmethod

class Command(ABC):
    @abstractmethod
    def execute(self): pass

    @abstractmethod
    def undo(self): pass

class TextEditor:
    def __init__(self):
        self.content = ""

    def add(self, text: str):
        self.content += text

    def remove(self, length: int):
        self.content = self.content[:-length]

    def get_content(self) -> str:
        return self.content

class TypeCommand(Command):
    def __init__(self, editor: TextEditor, text: str):
        self.editor = editor
        self.text = text

    def execute(self):
        self.editor.add(self.text)

    def undo(self):
        self.editor.remove(len(self.text))

# 利用例
editor = TextEditor()
history: list[Command] = []

cmd1 = TypeCommand(editor, "Hello ")
cmd1.execute()
history.append(cmd1)

cmd2 = TypeCommand(editor, "World")
cmd2.execute()
history.append(cmd2)

print(editor.get_content())  # Hello World

# Undo
last = history.pop()
last.undo()

print(editor.get_content())  # Hello 
